### PR TITLE
Issue 3899: Use external DNS name as segment store advertised endpoint

### LINKED ISSUE #3899: Use external DNS name as segment store advertised endpoint
**Problem description**
For TLS to work end-to-end with hostname verification enabled, the segment store's advertised endpoint must be a DNS name.   The advertised value is determined at segment store startup (`init_kubernetes.sh`) and has a problem; the external IP is advertised, rather than the external DNS name as required for hostname verification.

**Problem location**
`docker/pravega/scripts/init_kubernetes.sh`

**Suggestions for an improvement**
1. The external DNS name is available as an annotation on the `Service` resource.   Use that value in lieu of the IP address that is contained in the `status` section.  The annotation is:
`external-dns.alpha.kubernetes.io/hostname`
Note that the value may have a period as a suffix; if it does, strip it.  [More info about the annotation](https://github.com/kubernetes-incubator/external-dns/blob/master/docs/faq.md#how-do-i-specify-a-dns-name-for-my-kubernetes-objects).

### PR DESCRIPTION
**Pull request by pbelgundi**
**Change log description**  
Published address for segment store pods should be external dns name if that is provided via annotation: external-dns.alpha.kubernetes.io/hostname on the segment store external service.

**Purpose of the change**  
Fixes #3899 

**What the code does**  
If annotation 'external-dns.alpha.kubernetes.io/hostname' is set on segement store external service, the value against the annotation used as PUBLISHED_ADDRESS else ip address is used.

**How to verify it**  
Verified using pravega image with these changes and dns name was used as published address when domainName was specified in the manifest.
```
Connected to pravega-bookie-2.pravega-bookie-headless.default:3181
Running in a Kubernetes environment and managed by the Pravega Operator with external access enabled
Trying to obtain LoadBalancer external endpoint...
Found external service endpoint: pravega-pravega-segmentstore-0.example.com:12345
```

In the absence of domainName in manifest, IP is used:
```
Connected to pravega-bookie-2.pravega-bookie-headless.default:3181
Running in a Kubernetes environment and managed by the Pravega Operator with external access enabled
Trying to obtain LoadBalancer external endpoint...
Found external service endpoint: 10.240.119.149:12345
```

### CODE REVIEW

File: docker/pravega/scripts/init_kubernetes.sh [hunk 1]
```diff
@@ -47,17 +47,20 @@ init_kubernetes() {
...
         export PUBLISHED_PORT=""
-
+	
+	export PUBLISHED_ADDRESS=$( k8 "${ns}" "services" "${podname}" ".metadata.annotations[\"external-dns.alpha.kubernetes.io/hostname\"]" )
         service_type=$( k8 "${ns}" "services" "${podname}" ".spec.type" )
         if [ "${service_type}" == "LoadBalancer" ]; then
             while [ -z ${PUBLISHED_ADDRESS} ] || [ -z ${PUBLISHED_PORT} ]
             do
                 echo "Trying to obtain LoadBalancer external endpoint..."
                 sleep 10
```
Comment (EronWright) [inline docker/pravega/scripts/init_kubernetes.sh]: Looks to be that the script would sleep unnecessarily for at least 10 seconds in all cases.  The sleep is for retry purposes in the event that the status wasn't yet populated, which isn't relevant in the annotation case.   The fix would simply be to initialize the `PUBLISHED_PORT` where it is declared.

File: docker/pravega/scripts/init_kubernetes.sh [hunk 2]
```diff
@@ -47,17 +47,20 @@ init_kubernetes() {
         local podname=${POD_NAME}
         export PUBLISHED_ADDRESS=""
         export PUBLISHED_PORT=""
-
+	
+	export PUBLISHED_ADDRESS=$( k8 "${ns}" "services" "${podname}" ".metadata.annotations[\"external-dns.alpha.kubernetes.io/hostname\"]" )
```
Comment (EronWright) [inline docker/pravega/scripts/init_kubernetes.sh]: The trailing period must be stripped (if any).

Comment (pbelgundi) [reply docker/pravega/scripts/init_kubernetes.sh]: Done.

File: docker/pravega/scripts/init_kubernetes.sh [hunk 3]
```diff
@@ -47,28 +47,39 @@ init_kubernetes() {
         local podname=${POD_NAME}
         export PUBLISHED_ADDRESS=""
         export PUBLISHED_PORT=""
+	
+	      export PUBLISHED_ADDRESS=$( k8 "${ns}" "services" "${podname}" ".metadata.annotations[\"external-dns.alpha.kubernetes.io/hostname\"]" )
+
+        if [[ -n ${PUBLISHED_ADDRESS} && "${PUBLISHED_ADDRESS:${#PUBLISHED_ADDRESS}-1}" == "." ]];
+        then
+          export PUBLISHED_ADDRESS=${PUBLISHED_ADDRESS::-1}
```
[the remark this answers is not in the document]
Comment (shrids) [reply docker/pravega/scripts/init_kubernetes.sh]: https://github.com/pravega/pravega-operator/issues/211 ...

File: docker/pravega/scripts/init_kubernetes.sh [hunk 4]
```diff
@@ -47,28 +47,39 @@ init_kubernetes() {
         local podname=${POD_NAME}
         export PUBLISHED_ADDRESS=""
         export PUBLISHED_PORT=""
+	
+	      export PUBLISHED_ADDRESS=$( k8 "${ns}" "services" "${podname}" ".metadata.annotations[\"external-dns.alpha.kubernetes.io/hostname\"]" )
```
[the remark this answers is not in the document]
Comment (pbelgundi) [reply docker/pravega/scripts/init_kubernetes.sh]: Fixed

File: docker/pravega/scripts/init_kubernetes.sh (hunk 1 above)
Comment (pbelgundi) [reply docker/pravega/scripts/init_kubernetes.sh]: Fixed

**Pull request merged**